## Homework 2, FMNLP Fall 2023
### You have ten days for this assignment, please submit as PDF(File>Print>Save as PDF). 100 points total.

---
If you have any further questions or concerns, contact the TA via email:
baghbani.hamed@gmail.com or baghbani.hamed@ut.ac.ir
#####Telegram: @HBaghbani

***LINK: *paste your link here****

---

#### IMPORTANT: After copying this notebook to your Google Drive, please paste a link to it below. To get a publicly-accessible link, hit the *Share* button at the top right, then click "Get shareable link" and copy over the result. If you fail to do this, you will receive no credit for this homework!
***LINK:*** https://colab.research.google.com/drive/1ZyzWqyAsrX0qp-x6CAAHmSyYHphImdSD?usp=sharing

---
##### *How to submit this problem set:*
- Write all the answers in this Colab notebook. Once you are finished, generate a PDF via (File -> Print -> Save as PDF) and upload it to Elearn (elearn.ut.ac.ir).
  
- **Important:** check your PDF before you submit to Elearn to make sure it exported correctly. If Colab gets confused about your syntax, it will sometimes terminate the PDF creation routine early.

- **Important:** on Elearn, please make sure that you tag each page with the corresponding question(s). This makes it significantly easier for our graders to grade submissions, especially with the long outputs of many of these cells. We will take off points for submissions that are not tagged.

- When creating your final version of the PDF to hand in, please do a fresh restart and execute every cell in order. One handy way to do this is by clicking `Runtime -> Run All` in the notebook menu.

---

##### *Academic honesty*

- We will audit the Colab notebooks from a set number of students, chosen at random. The audits will check that the code you wrote actually generates the answers in your PDF. If you turn in correct answers on your PDF without code that actually generates those answers, we will consider this a serious case of cheating. See the course page for honesty policies.

- We will also run automatic checks of Colab notebooks for plagiarism. Copying code from others is also considered a serious case of cheating.

---

# Part 1: Annotation

In this homework, you will first collect a labeled dataset of 360 **Persian sentences**  for a text classification task of your choice.
You can do this in teams of size three.
(Note that you can build teams of one or two as well but then you need to ask two or one other annotators to help you.)
The process will include:

1. *Data collection*: Collect 360 Persian sentences from any source you find interesting (e.g., literature, Tweets, news articles, reviews, etc.)

2.  *Task design*: Come up with a binary (i.e., only two labels) sentence-level classification task that you would like to perform on your sentences. Be creative, and make sure your task isn't too easy (e.g., perhaps the labels have some degree of subjectivity to them, or the task is otherwise complex for humans). Write up annotator guidelines/instructions on how you would like people to label your data

3. On your dataset, every example must be annotated by three persons. In order to get everything done on time, you need to complete the following steps:

>  * Build annotation guideline
*   Annotate the data (e.g., a spreadsheet or Google form)
*   Collect the labeled data from each of the three annotators.
*   Sanity check the data for basic cleanliness (are all examples annotated? are all labels allowable ones?)

4. Collect feedback from all the team members about the task including annotation time and obstacles encountered (e.g., maybe your guidelines were confusing! or maybe some sentences were particularly hard to annotate!)

5. Calculate and report inter-annotator agreement.

6. Aggregate output from three annotators to create final dataset.

7. Perform NLP experiments on your new dataset!

## Question 1.1 (10 points):
Describe the source of your unlabeled data, why you chose it, and what kind of sentence selection process you used (if any) to choose 120 sentences for annotation. Also briefly describe the text classification task that you will be collecting labels for in the next section.

### *WRITE YOUR ANSWER HERE* ###

Since I consider myself an avid gamer, after multiple discussions with my teammates, we decided to use the comments and review/comment section of the [Hayoola website](https://hayoola.com/#!/home). This website is in Persian. In this website the game developers publish their developed games online. People pay real money to buy the games, and get to play it. This website can be considered an Iranian version of [Steam website](https://store.steampowered.com/).

Some people, after finishing a game, write a review/comment on that game, or leave a comment about it. Their points on how the games were good, or bad, or chould have been better, can be useful. We wondered that this type of dataset would never see the light of the day, since, Iranian games are kind of in minority in our country, and not that many gamers play them (compared the foriegn games being played in our country). So, we seized this opportunity as to introduce a new source or dataset of Persian text, other than Instagram, Twitter or regular news websites.

We crawled the said website using a Python source code, and collected reviews/comments and comments of the people on games. We selected only Persian text, since, there were some Finglish comments as well. We tried to be diverse, inclusive of most genres and games. Our gathered reviews/comments are of any size, whether long or short, or even an essay-like comprising of multiple sentences. Sentiment-wise, we tried our absolute hardest to collect as balanced dataset as we could. Overall, our collected samples can be considered a good representative and also an introduction to this website for further data collection process.

After we collected our dataset and our decisions on what sentences to use and what to ignore, we decided to label all sentences independently, without knowing others' label. We used binary labeling system: 1 means the overall sentiment of the review/comment is positive; 0 means the overall sentiment is negative. Each of us considered the review/comment as a whole, and decided on the overall intent of the user.

One positive note on our gathered dataset can be attributed to the fact that the reviews/comments in our dataset vary in size. Some reviews/comments are really long, comprising of multiple sentences, in which, the user has tried their best to describe the positive and negative aspects of the game. Some reviews/comments are short and in them, the user has just stated their enjoyment or disappointment of the game. We had this goal in our mind that we like to fine-tune the pre-trained model on a diverse set of reviews/comments, both in sentiment and size of the review/comment. When we give the model really long reviews/comments, we like the model to understand the whole intent of the user, rather than fixating on just few short sentences. In this way, we can be sure that if the model has performed well, it was because it really understood the intent of the review/comment, rather than just fixating on some few keywords. Thus, the model has to be general and robust at the same time, and try its best to look at a review/comment as a whole and understand the user's intent, rather than memorization. Because, some reviews/comments may use bad points or strong criticism against a specific game, but they mean well and have enjoyed the game a lot, and want the game to be more successful.

## Question 1.2 (25 points):
Copy the annotation guidelines that you provided to your classmates below. We expect that these guidelines will be very detailed (and as such could be fairly long). You must include:

> *   The two categories for your binary classification problem, including the exact strings you want annotators to use while labeling.
*   Descriptions of the categories and what they mean.
*   Representative examples of each category (i.e., sentences from outside your dataset that you have manually labeled to give annotators an idea of how to perform the task)
*   A discussion of of tricky corner cases, and criteria to help the annotator decide them. If you look at the data and think about how an annotator could do the task, you will likely find a bunch of these!


---




### *COPY YOUR ANNOTATION GUIDELINES HERE.* Please format them nicely so it's easy for us to read / grade :)###

The overall guidelines I had on my mind when labeling the reviews/comments and also what I told my teammates are as follow:
* I only considered the author's intent as to only be positive or negative, ignoring other labels such as neutral.
* A review/comment might contain negative wording or sentences. I considered the overall intent of the author. For example, a review/comment might contain some criticism, but, overall the author has enjoyed the game a lot and the criticism is just for letting the game developers know how to improve the game. Overall, I determined the overall sentiment of the review/comment, whether it leans towards a positive or negative evaluation of the game, not letting minor details or criticism points have more weight.
* I mostly considered the user's subjective opinion, and the overall emotions they felt during or after they wrote a review/comment; and also, I considered objective side of the review/comment, what the user has felt and meant to tell. I balanced between these two, and labeled the review/comment accordingly based upon what I thought of the review/comment, and overall experience the user has felt.
* I did not consider the number of likes or dislikes a review has gotten. I only read the text.
* I familiarized myself with the specific context of the game being reviewed. This helped me understand the user's perspective and the elements they focused on.
* I excluded objective information such as game specifications, facts, or technical details that do not express the user's sentiment. As I said above, I only focused on the emotional content of the reviews/comments.
* I paid attention to any key indicators of sentiment including language tone, specific adjectives, adverbs, and emotional expressions that suggest positivity or negativity.
* It is again important to mention that we all reviewed all the reviews/comments once, and labeled it independently without letting the other teammates know of our labels.
* And last but not least, I ensured that my annotations are not influenced by my personal opinions. I stayed objective and focused on the user's perspective.

In [ ]:
# the piece of code we wrote to collect and gather reviews/comments from Hayoola website:

import time
import json
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.wait import WebDriverWait

driver = webdriver.Chrome()
driver.get("https://hayoola.com/games#!/search/0")

# some util functions to makes our code cleaner & more reliable:
def get_name() -> str:
    return driver.find_element(By.CLASS_NAME, "game-title").get_attribute("innerText")

def get_score() -> float:
    return float(driver.find_element(By.CLASS_NAME, "game-rating-value").get_attribute("innerText").split()[0])

def get_rating() -> float:
    return float(driver.find_element(By.CLASS_NAME, "game-comment-count").get_attribute("innerText").split()[0])

def get_reviews() -> list:
    reviews = []

    for _ in range(10):
        driver.execute_script("window.scrollTo(0, document.body.scrollHeight)")
        time.sleep(1)

    for review in driver.find_elements(By.CLASS_NAME, "game-review"):
        date = review.find_element(By.CLASS_NAME, "game-review-date").get_attribute("innerText")

        rate = (review.find_element(By.CLASS_NAME, "game-rating-value").get_attribute("innerText").split()[0])
        text = review.find_element(By.CLASS_NAME, "limitted-height-el").get_attribute("innerText")

        votes = tuple()

        try:
            votes = (review.find_element(By.CLASS_NAME, "game-review-vote-details").get_attribute("innerText").split())
            votes = tuple(map(int, filter(lambda x: str(x).isdigit(), votes)))
        except:
            pass

        reviews.append({"date": date, "rate": float(rate), "text": text, "votes": votes})

    return reviews

# the following loop extracts only the game links:

game_links = dict()

for p in range(4):
    WebDriverWait(driver, 1)

    game_links[f"page_{p + 1}"] = [x.get_attribute("href") for x in driver.find_elements(By.CLASS_NAME, "game-card")]

    try:
        driver.find_element(By.XPATH, f"//button[contains(text(), '{p + 2}')]").click()
        time.sleep(5)
        WebDriverWait(driver, 2)
    except:
        break

# the following loop extracts data from each game's page:

games = []

for link in set(sum(game_links.values(), [])):
    try:
        driver.get(link)
        WebDriverWait(driver, timeout=2)

        games.append({"link": link, "name": get_name(), "score": get_score(), "ratings": get_rating(), "reviews": get_reviews()})

        print("Total Reviews:", sum(map(lambda x: len(x["reviews"]), games)))
    except KeyboardInterrupt:
        print("DONE")
        break

with open("games.json", "w") as f:
    json.dump(games, f, indent=4, ensure_ascii=False)

with open("gameLinks.json", "w") as f:
    json.dump(game_links, f, indent=4, ensure_ascii=False)

In [ ]:
# the piece of code i used to read the original data we collected:

import json
import pandas as pd

with open("games.json", encoding="utf-8") as file:
    data = json.load(file)

reviews_text = [review["text"] for item in data for review in item["reviews"]]

df = pd.DataFrame(columns=["text", "label"])

df["text"] = reviews_text
print(df.head())

#df.to_csv("reviews with no labels.csv", index=None)

In [ ]:
# the piece of code i utilized for labelling:

import pandas as pd

df = pd.read_csv("reviews with no labels.csv")

df["label"] = -1666
print(df.head())

for i in range(df.shape[0]):
    if df.loc[i, "label"] == -1666:
        print(df.loc[i, "text"])

        df.loc[i, "label"] = int(input("\nenter your label: "))

#df.to_csv("1.csv", index=None)

## Question 1.3 (5 points):
Write down the names and emails of the two classmates who will be annotating your data below. Once they are finished annotating, create two .csv files (annotator1.csv and annotator2.csv) that contains each annotator's labels. The file should have two columns with headers **text** and **label**, respectively. You will include these files in an email to the instructors account when you're finished with this homework.

*The tweets.csv file provided as an example in Part 2 below uses the same format.*

### *WRITE CLASSMATE 1 NAME/EMAIL HERE:* ### Mohammad Mehdi Mohajeri (mehdimohajeri78@gmail.com)
### *WRITE CLASSMATE 2 NAME/EMAIL HERE:* ### Mahyar Riazati (mr.riazati1999@gmail.com)

## Question 1.4 (10 points):
After both annotators have finished labeling the 120 sentences you gave them, ask them for feedback about your task and the provided annotation guidelines. If you were to collect more labeled data for this task in the future, what would you change from your current setup? Why? Please include a summary of annotator feedback (with specific examples that they found challenging to label) in your answer.

### *WRITE ANSWER HERE* ###

After each of us finished labeling the reviews/comments, we talked about the process and experience each of us had while labeling. Some reviews/comments had conflicting points in the whole paragraph or sentences, but the overall intent was positive. However, we were on the fence as to whether label these types of reviews/comments as positive or negative. Those who supported the idea of the review/comment being positive, claimed that the constructive criticism should not be considered negative; the most important thing is that the user has enjoyed the game. However, on the other side, those who supported the negative labeling, pointed to the multiple issues a user for example has had, and this fact has rendered the users' experience down; thus, labeling the review/comment as negative, even though the user said the game is good. We discussed this topic for multiple sessions and we heard the opposing opinions and feedbacks.

As said previously, we considered both objective and emotional aspects of a review. We balanced between these two, and labeled the review/comment accordingly, based upon what we think of the review/comment and overall experience.

Another issue we all encountered was that there could have been another label, such as neutral. Some reviews/comments were not solely about reviewing the game or saying one's opinion about the game. Rather, they were about thanking the developers of a game, asking questions, or bringing topics and issues which were not relevant to the game at all. In these types of reviews/comments, we considered the tone of the user, whether they tried to be genuine or sarcastic, and we labeled accordingly.

And finally, if we were to collect more labeled data for this task in the future, we would consider collecting reviews/comments from other similar websites as well. Iranian gamers who play Iranian games are not that many, however, from this experience that we had with Hayoola website, we believe any similar can be a great source of reviews/comments. Some gamers tend to review a game, or complain or praise it in great detail. Thus, the websites like this can be great.

## Question 1.5 (10 points):
Now, compute the inter-annotator agreement between your two annotators. Upload both .csv files to your Colab session (click the folder icon in the sidebar to the left of the screen). In the code cell below, read the data from the two files and compute both the raw agreement (% of examples for which both annotators agreed on the label) and the [Cohen's Kappa](https://en.wikipedia.org/wiki/Cohen%27s_kappa). Feel free to use implementations in existing libraries (e.g., [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.cohen_kappa_score.html)). After you're done, paste the numbers in the text cell that follows your code.

*If you're curious, Cohen suggested the Kappa result be interpreted as follows: values ≤ 0 as indicating no agreement and 0.01–0.20 as none to slight, 0.21–0.40 as fair, 0.41– 0.60 as moderate, 0.61–0.80 as substantial, and 0.81–1.00 as almost perfect agreement.*

In [ ]:
### WRITE CODE TO LOAD ANNOTATIONS AND
### COMPUTE AGREEMENT + COHEN'S KAPPA HERE!

import pandas as pd
from sklearn.metrics import cohen_kappa_score

y_mahyar = pd.read_csv("mahyar.csv")["label"]
y_emad = pd.read_csv("emad.csv")["label"] # we call Mohammad Mehdi, Emad .)))

cohen_kappa = cohen_kappa_score(y_mahyar, y_emad)
print(f"COHEN'S KAPPA: {cohen_kappa:.5f}")

raw_agreement = sum(1 for annotation1, annotation2 in zip(y_mahyar, y_emad) if annotation1 == annotation2) / len(y_mahyar) * 100
print(f"RAW AGREEMENT: {raw_agreement:.2f}%")

COHEN'S KAPPA: 0.41968
RAW AGREEMENT: 76.88%


###*RAW AGREEMENT*: 76.88%
###*COHEN'S KAPPA*: 0.41968

## Question 1.6 (10 points):
To form your final dataset, you need to *aggregate* the annotations from both annotators (i.e., for cases where they disagree, you need to choose a single label). Use any method you like other than random label selection to perform this aggregation (e.g., have the two annotators discuss each disagreement and come to consensus, or choose the label you agree with the most). Upload your final dataset to the Colab session (in the same format as the other two files) as final_dataset.csv. Remember to include this file in your final submit to us!

### *DESCRIBE YOUR AGGREGATION STRATEGY HERE* ###

After we discussed about how we performed labeling for reviews/comments, I had two approaches to decide between labels: relabeling my label for the reviews/comments based upon my teammates' votes; or do a majority-voting strategy, in which, the label with most occurence would be the final label.

I have written the code for both parts. However, the approach I chose was the first one. After some discussions and hearing my teammates' opinions and seeing their labeling strategy and also labels, I decided to relabel the reviews/comments, with having their mindset alongisde mine as well. For this, I sat down and checked any review/comment which were labeled differently by us, and we did not label them the same. I read those reviews/comments again and considered the labels, my teammates had assigned to them. Then, I decided a suitable label for each review/comment. Sometimes, I stood my ground and used the label i previously gave to that review/comment. And sometimes, I used one of teammate's label for that review. I put myself in my teammates' shoes and reviewed the review/comment from their perspective. I chose this approach as my main.

Another approach, which I also implemented below, is majority-voting between votes of all three of us. I also considered this approach before my final decision on which approach to choose, since, I reasoned that maybe it is better to choose the label which has more votes on it. This approach can be fair and objective sometimes, but it cannot be more accurate and better than the approach above, at least in my opinion. Since, some reviews/comments needed thorough inspection. Therefore, I did not choose this approach.

In [ ]:
# the piece of code i utilized for my main approach:

import pandas as pd

df_emad = pd.read_csv("emad.csv")
df_mahyar = pd.read_csv("mahyar.csv")
df_1 = pd.read_csv("1.csv")

df = pd.DataFrame(columns=["text", "label"])

df["text"] = df_1["text"]
df["label"] = -1666

for i in range(df.shape[0]):
    if df_emad.loc[i, "label"] + df_mahyar.loc[i, "label"] + df_1.loc[i, "label"] == 0 or \
    df_emad.loc[i, "label"] + df_mahyar.loc[i, "label"] + df_1.loc[i, "label"] == 3:
        df.loc[i, "label"] = df_1.loc[i, "label"]
    else:
        print(df.loc[i, "text"])

        print("\nMAHYAR: ", df_mahyar.loc[i, "label"])
        print("EMAD: ", df_emad.loc[i, "label"])
        print("1: ", df_1.loc[i, "label"])

        df.loc[i, "label"] = int(input("\nenter your label: "))

display(df)

#df.to_csv("final_data.csv", index=None)

In [ ]:
# another approach to aggregate annotations from all annotators:

df_emad = pd.read_csv("emad.csv")
df_mahyar = pd.read_csv("mahyar.csv")
df_1 = pd.read_csv("1.csv")

final_data = pd.DataFrame(columns=["text", "label"])

final_data["text"] = df_1["text"]
final_data["label_1"] = df_1["label"]
final_data["label_emad"] = df_emad["label"]
final_data["label_mahyar"] = df_mahyar["label"]

# inspired by https://stackoverflow.com/a/55749310
final_data["label"] = final_data.mode(axis=1)

final_data.drop(columns=["label_1", "label_emad", "label_mahyar"], inplace=True)

display(final_data)

#final_data.to_csv("final_data.csv", index=None)

text  label
0    بازی خوبی بود ولی چون چک پوینت کم داشت و با وج...      0
1    بازی بسیار زیبا در عین حال با موسیقی های 8 بیت...      1
2    سلام بازی خیلی خوبیه هر چند یه سری ایراد داره ...      1
3               واقعا دمتون گرم با این بازی خیلی عالیه      1
4    بازی قشنگ بود اما تورو جدتون دیالوگ ها رو کتاب...      0
..                                                 ...    ...
380  این بازی ایرانی به شدت توصیه می شود. نقاط قوت:...      1
381           خوب بود ولی تو یه مرحله گیر کرد راه نرفت      1
382  بازی در بیشتر مورد هایی که من می خواستم خوب بو...      1
383  بینظیر بهترین بازی که برای تاریخ کشورمون هست ا...      1
384  خیلی خوبه وقتی بازی میکنی انگار واقعا داری میج...      1

[385 rows x 2 columns]

# Part 2: Text classification

Now we'll move onto fine-tuning  pretrained language models specifically on your dataset. This part of the homework is meant to be an introduction to the HuggingFace library, and it contains code that will potentially be useful for your final projects. Since we're dealing with large models, the first step is to change to a GPU runtime.

## Adding a hardware accelerator

Please go to the menu and add a GPU as follows:

`Edit > Notebook Settings > Hardware accelerator > (GPU)`

Run the following cell to confirm that the GPU is detected.

In [ ]:
import torch

# Confirm that the GPU is detected

assert torch.cuda.is_available()

# Get the GPU device name.
device_name = torch.cuda.get_device_name()
n_gpu = torch.cuda.device_count()
print(f"Found device: {device_name}, n_gpu: {n_gpu}")
device = torch.device("cuda")

Found device: Tesla T4, n_gpu: 1


## Installing Hugging Face's Transformers library
We will use Hugging Face's Transformers (https://github.com/huggingface/transformers), an open-source library that provides general-purpose architectures for natural language understanding and generation with a collection of various pretrained models made by the NLP community. This library will allow us to easily use pretrained models like `BERT` and perform experiments on top of them. We can use these models to solve downstream target tasks, such as text classification, question answering, and sequence labeling.

Run the following cell to install Hugging Face's Transformers library and download a sample data file called tweets.csv that contains tweets about airlines along with a negative, neutral, or positive sentiment rating. Note that you will be asked to link with your Google Drive account to download some of these files. If you're concerned about security risks (there have not been any issues in previous semesters), feel free to make a new Google account and use it for this homework!

In [ ]:
!pip install transformers
!pip install -U -q PyDrive

from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials
# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)
print('success!')

import os
import zipfile

# Download helper functions file
helper_file = drive.CreateFile({'id': '16HW-z9Y1tM3gZ_vFpJAuwUDohz91Aac-'})
helper_file.GetContentFile('helpers.py')
print('helper file downloaded! (helpers.py)')

# Download sample file of tweets
data_file = drive.CreateFile({'id': '1QcoAmjOYRtsMX7njjQTYooIbJHPc6Ese'})
data_file.GetContentFile('tweets.csv')
print('sample tweets downloaded! (tweets.csv)')

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.9/7.9 MB 13.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 311.2/311.2 kB 29.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 31.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 38.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.0/295.0 kB 31.4 MB/s eta 0:00:00
success!
helper file downloaded! (helpers.py)
sample tweets downloaded! (tweets.csv)


The cell below imports some helper functions we wrote to demonstrate the task on the sample tweet dataset.

In [ ]:
from helpers import tokenize_and_format, flat_accuracy

# Part 1: Data Prep and Model Specifications

Upload your data using the file explorer to the left. We have provided a function below to tokenize and format your data as BERT requires. Make sure that your csv file, titled final_data.csv, has one column "text" and another column "labels" containing integers.

If you run the cell below without modifications, it will run on the tweets.csv example data we have provided. It imports some helper functions we wrote to demonstrate the task on the sample tweet dataset. You should first run all of the following cells with tweets.csv just to see how everything works. Then, once you understand the whole preprocessing / fine-tuning process, change the csv in the below cell to your final_data.csv file, add any extra preprocessing code you wish, and then run the cells again on your own data.

In [ ]:
from helpers import tokenize_and_format, flat_accuracy
import pandas as pd

df = pd.read_csv('final_data.csv')

df = df.sample(frac=1).reset_index(drop=True)

texts = df.text.values
labels = df.label.values

### tokenize_and_format() is a helper function provided in helpers.py ###
input_ids, attention_masks = tokenize_and_format(texts)

# Convert the lists into tensors.
input_ids = torch.cat(input_ids, dim=0)
attention_masks = torch.cat(attention_masks, dim=0)
labels = torch.tensor(labels)

# Print sentence 0, now as a list of IDs.
print('Original: ', texts[0])
print('Token IDs:', input_ids[0])

Original:  لذت بردم و دارم میبرم واقعا بی نظیر بود و همچنان دارم لذت میبرم ازش
Token IDs: tensor([  101,  1294, 29822, 29817,  1271, 17149, 15394, 22192,  1298,  1278,
        25573, 17149, 22192,  1295, 24830, 29816, 17149, 22192,  1298, 25573,
        29834, 29830, 25573,  1271, 24830,  1296, 29829, 24830, 17149,  1271,
        29836, 15394,  1298,  1297, 22192, 29840, 15915, 18511,  1278, 25573,
        17149, 22192,  1294, 29822, 29817,  1295, 24830, 29816, 17149, 22192,
         1270, 29823, 29825,   102,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0])


## Create train/test/validation splits

Here we split your dataset into 3 parts: a training set, a validation set, and a testing set. Each item in your dataset will be a 3-tuple containing an input_id tensor, an attention_mask tensor, and a label tensor.



In [ ]:

total = len(df)

num_train = int(total * .8)
num_val = int(total * .1)
num_test = total - num_train - num_val

# make lists of 3-tuples (already shuffled the dataframe in cell above)

train_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_train)]
val_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_train, num_val+num_train)]
test_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_val + num_train, total)]

train_text = [texts[i] for i in range(num_train)]
val_text = [texts[i] for i in range(num_train, num_val+num_train)]
test_text = [texts[i] for i in range(num_val + num_train, total)]


Here we choose the model we want to finetune from https://huggingface.co/transformers/pretrained_models.html. Because the task requires us to label sentences, we wil be using BertForSequenceClassification below. You may see a warning that states that `some weights of the model checkpoint at [model name] were not used when initializing. . .` This warning is expected and means that you should fine-tune your pre-trained model before using it on your downstream task. See [here](https://github.com/huggingface/transformers/issues/5421#issuecomment-652582854) for more info.

In [ ]:
from transformers import BertForSequenceClassification, AdamW, BertConfig

model = BertForSequenceClassification.from_pretrained(
    "bert-base-uncased", # Use the 12-layer BERT model, with an uncased vocab.
    num_labels = 2, # The number of output labels.
    output_attentions = False, # Whether the model returns attentions weights.
    output_hidden_states = False, # Whether the model returns all hidden-states.
)

# Tell pytorch to run this model on the GPU.
model.cuda()


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


BertForSequenceClassification(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12,

# ACTION REQUIRED #

Define your fine-tuning hyperparameters in the cell below (we have randomly picked some values to start with). We want you to experiment with different configurations to find the one that works best (i.e., highest accuracy) on your validation set. Feel free to also change pretrained models to others available in the HuggingFace library (you'll have to modify the cell above to do this). You might find papers on BERT fine-tuning stability (e.g., [Mosbach et al., ICLR 2021](https://openreview.net/pdf?id=nzpLWnVAyah)) to be of interest.

In [ ]:
batch_size = 128
optimizer = AdamW(model.parameters(),
                  lr = 5e-5, # args.learning_rate - default is 5e-5
                  eps = 1e-8 # args.adam_epsilon  - default is 1e-8
                )
epochs = 20

/usr/local/lib/python3.10/dist-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


# Fine-tune your model
Here we provide code for fine-tuning your model, monitoring the loss, and checking your validation accuracy. Rerun both of the below cells when you change your hyperparameters above.

In [ ]:
import numpy as np
# function to get validation accuracy
def get_validation_performance(val_set):
    # Put the model in evaluation mode
    model.eval()

    # Tracking variables
    total_eval_accuracy = 0
    total_eval_loss = 0

    num_batches = int(len(val_set)/batch_size) + 1

    total_correct = 0

    for i in range(num_batches):

      end_index = min(batch_size * (i+1), len(val_set))

      batch = val_set[i*batch_size:end_index]

      if len(batch) == 0: continue

      input_id_tensors = torch.stack([data[0] for data in batch])
      input_mask_tensors = torch.stack([data[1] for data in batch])
      label_tensors = torch.stack([data[2] for data in batch])

      # Move tensors to the GPU
      b_input_ids = input_id_tensors.to(device)
      b_input_mask = input_mask_tensors.to(device)
      b_labels = label_tensors.to(device)

      # Tell pytorch not to bother with constructing the compute graph during
      # the forward pass, since this is only needed for backprop (training).
      with torch.no_grad():

        # Forward pass, calculate logit predictions.
        outputs = model(b_input_ids,
                                token_type_ids=None,
                                attention_mask=b_input_mask,
                                labels=b_labels)
        loss = outputs.loss
        logits = outputs.logits

        # Accumulate the validation loss.
        total_eval_loss += loss.item()

        # Move logits and labels to CPU
        logits = logits.detach().cpu().numpy()
        label_ids = b_labels.to('cpu').numpy()

        # Calculate the number of correctly labeled examples in batch
        pred_flat = np.argmax(logits, axis=1).flatten()
        labels_flat = label_ids.flatten()
        num_correct = np.sum(pred_flat == labels_flat)
        total_correct += num_correct

    # Report the final accuracy for this validation run.
    avg_val_accuracy = total_correct / len(val_set)
    return avg_val_accuracy



In [ ]:
import random

# training loop

# For each epoch...
for epoch_i in range(0, epochs):
    # Perform one full pass over the training set.

    print("")
    print('======== Epoch {:} / {:} ========'.format(epoch_i + 1, epochs))
    print('Training...')

    # Reset the total loss for this epoch.
    total_train_loss = 0

    # Put the model into training mode.
    model.train()

    # For each batch of training data...
    num_batches = int(len(train_set)/batch_size) + 1

    for i in range(num_batches):
      end_index = min(batch_size * (i+1), len(train_set))

      batch = train_set[i*batch_size:end_index]

      if len(batch) == 0: continue

      input_id_tensors = torch.stack([data[0] for data in batch])
      input_mask_tensors = torch.stack([data[1] for data in batch])
      label_tensors = torch.stack([data[2] for data in batch])

      # Move tensors to the GPU
      b_input_ids = input_id_tensors.to(device)
      b_input_mask = input_mask_tensors.to(device)
      b_labels = label_tensors.to(device)

      # Clear the previously calculated gradient
      model.zero_grad()

      # Perform a forward pass (evaluate the model on this training batch).
      outputs = model(b_input_ids,
                            token_type_ids=None,
                            attention_mask=b_input_mask,
                            labels=b_labels)
      loss = outputs.loss
      logits = outputs.logits

      total_train_loss += loss.item()

      # Perform a backward pass to calculate the gradients.
      loss.backward()

      # Update parameters and take a step using the computed gradient.
      optimizer.step()

    # ========================================
    #               Validation
    # ========================================
    # After the completion of each training epoch, measure our performance on
    # our validation set. Implement this function in the cell above.
    print(f"Total loss: {total_train_loss}")
    val_acc = get_validation_performance(val_set)
    print(f"Validation accuracy: {val_acc}")

print("")
print("Training complete!")



======== Epoch 1 / 20 ========
Training...
Total loss: 1.710372805595398
Validation accuracy: 0.7631578947368421

======== Epoch 2 / 20 ========
Training...
Total loss: 1.698032796382904
Validation accuracy: 0.7631578947368421

======== Epoch 3 / 20 ========
Training...
Total loss: 1.719005286693573
Validation accuracy: 0.7631578947368421

======== Epoch 4 / 20 ========
Training...
Total loss: 1.713633954524994
Validation accuracy: 0.7631578947368421

======== Epoch 5 / 20 ========
Training...
Total loss: 1.7040157914161682
Validation accuracy: 0.7631578947368421

======== Epoch 6 / 20 ========
Training...
Total loss: 1.7388708591461182
Validation accuracy: 0.7631578947368421

======== Epoch 7 / 20 ========
Training...
Total loss: 1.7366893291473389
Validation accuracy: 0.7631578947368421

======== Epoch 8 / 20 ========
Training...
Total loss: 1.7219849824905396
Validation accuracy: 0.7631578947368421

======== Epoch 9 / 20 ========
Training...
Total loss: 1.741681694984436
Validation

# Evaluate your model on the test set
After you're satisfied with your hyperparameters (i.e., you're unable to achieve higher validation accuracy by modifying them further), it's time to evaluate your model on the test set! Run the below cell to compute test set accuracy.


In [ ]:
get_validation_performance(test_set)

0.7692307692307693

## Question 2.1 (10 points):
Congratulations! You've now gone through the entire fine-tuning process and created a model for your downstream task. Two more questions left :) First, describe your hyperparameter selection process in words. If you based your process on any research papers or websites, please reference them. Why do you think the hyperparameters you ended up choosing worked better than others? Also, is there a significant discrepancy between your test and validation accuracy? Why do you think this is the case?

### *WRITE YOUR ANSWER HERE*

First question:

I tried many values for each of the hyperparameters, mostly learning rate, epochs and batch size.

The BERT paper recommended either 32 or 64 for batch size. Aside from these values, I gave batch size the value of 256. However, sometimes Google Colab would send me error, telling me it has run out of memory. So, I degraded the batch size to 128. This amount is however, good and works well for me, so, I went with it.

In next step, for epochs I tried values such as 10, 20, 30, 50, 66, and 80. The more I trained the model, the stable the results were. Meaning, increasing the number of epochs did not have any positive effect for me. The best value I came up with was 20. I compared the results when I trained the model for 20 and 30 epochs, and, between these two I could not distinguish anything, and sometimes the performance was the same. So, I chose 20 for epochs.

At last, I tried multiple values such as 1e-3, 1e-5, 2e-3, 2e-5, 3e-5, 4e-5, 5e-5, 6e-5, 7e-5. Among these, only the values 5e-5 and 6e-5 worked really well and the loss was decreasing faster. When I used 6e-5 I got really different results for test set, from 53% to 82%. But with 5e-5, I got much more stable results. I do not understand why, and I searched small searches to understand the reason. Unfortunately, I could not find anything useful to explain as to why the model behaved the way it did.

I tried only two values for epsilon: 1e-8 and 2e-8. With 2e-8, I got worse results. So, I did not experiment on this hyperparameter that much and preferred to leave it as it is.

As to why I think the chosen hyperparameters worked for me, I have to say that I chose the number of epochs wisely. The more the model was being trained, the more stale or even worse, it performed. I preferred to use a wise number of epochs, so that the model would not fine-tune for wrong reasons and overfit my small dataset, therefore, leading to catastrophic forgetting of the model (as it was discussed in class). For learning rate, I went with the default value, because it was mostly stable and it performed a little bit better than other values. However, since, my dataset is small, I could have stayed with the 6e-5 value or even a little bit higher than 5e-5, if I had not seen conflicting results. Lastly, for batch size, I wanted to give the hyperparameter bigger value, since I have a small dataset and I could train the model on all of it at once (in fewer iterations inside each epoch), and also the model has trained on large batches. Larger values for batch size helped the model decrease its loss fast and this was reassuring for me to go higher than 128. However, since, Google Colab was unstable when I assigned the batch size to 256, I remained on 128. For epislon, I did not delve too much into experiments like other hyperparameters. However, I tried some values for it after reading about it and how it works or affects the training process. At the end, I preferred to choose the default value for this hyperparameter.

As for second question that if there is any discrepancy between test and validation accuracy, I have to say that I sometimes noticed discrepancy in my many and multiple experiments. However, it was not that acute, and I told myself the small size of test-set and validation-set can be a major factor for this. However, for the results I have provided for you, there is actually no discrepancy between the values of accuracy for test-set and validation-set. Since, I sometimes saw discrepancy and sometimes did not, I can add another factor such as learning rate, and how it has affected the performance of the model while training.

Overall, since, the reviews/comments in our dataset are diverse and the dataset is not overall imbalanced, I can be certain that if there was any discrepancy between the values of accuracy between these two sets, it can be attributed to how the model has learnt and performed on training-set.

For this question, I also did some research and read some papers. The values I brought above can either be referenced to these papers, or be determined by sheer trial-and-error done by me. I tried and compared the results for many possible combinations of values for each of the hyperparameters. Therefore, I reached to the final values for the hyperparameters.

Nevertheless, of the research papers I glanced at, I can name:

* [BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/abs/1810.04805)
* [Large Batch Optimization for Deep Learning: Training BERT in 76 minutes](https://arxiv.org/abs/1904.00962)
* [How to Fine-Tune BERT for Text Classification?](https://link.springer.com/chapter/10.1007/978-3-030-32381-3_16)
* [How to Train BERT with an Academic Budget](https://arxiv.org/abs/2104.07705)
* [Visualizing and Understanding the Effectiveness of BERT](https://arxiv.org/abs/1908.05620)
* [A Robustly Optimized BERT Pre-training Approach with Post-training](https://link.springer.com/chapter/10.1007/978-3-030-84186-7_31)

## Question 2.2 (20 points):
Finally, perform an *error analysis* on your model. This is good practice for your final project. Write some code in the below code cell to print out the text of up to five test set examples that your model gets **wrong**. If your model gets more than five test examples wrong, randomly choose five of them to analyze. If your model gets fewer than five examples wrong, please design five test examples that fool your model (i.e., *adversarial examples*). Then, in the following text cell, perform a qualitative analysis of these examples. See if you can figure out any reasons for errors that you observe, or if you have any informed guesses (e.g., common linguistic properties of these particular examples). Does this analysis suggest any possible future steps to improve your classifier?

In [ ]:
## YOUR ERROR ANALYSIS CODE HERE
## print out up to 5 test set examples (or adversarial examples) that your model gets wrong

def mis_classified_examples(val_set):
    # Put the model in evaluation mode
    model.eval()

    num_batches = int(len(val_set)/batch_size) + 1

    for i in range(num_batches):

      end_index = min(batch_size * (i+1), len(val_set))

      batch = val_set[i*batch_size:end_index]

      if len(batch) == 0: continue

      input_id_tensors = torch.stack([data[0] for data in batch])
      input_mask_tensors = torch.stack([data[1] for data in batch])
      label_tensors = torch.stack([data[2] for data in batch])

      # Move tensors to the GPU
      b_input_ids = input_id_tensors.to(device)
      b_input_mask = input_mask_tensors.to(device)
      b_labels = label_tensors.to(device)

      # Tell pytorch not to bother with constructing the compute graph during
      # the forward pass, since this is only needed for backprop (training).
      with torch.no_grad():

        # Forward pass, calculate logit predictions.
        outputs = model(b_input_ids,
                                token_type_ids=None,
                                attention_mask=b_input_mask,
                                labels=b_labels)

        logits = outputs.logits

        # Move logits and labels to CPU
        logits = logits.detach().cpu().numpy()
        label_ids = b_labels.to('cpu').numpy()

        # Calculate the number of correctly labeled examples in batch
        pred_flat = np.argmax(logits, axis=1).flatten()
        labels_flat = label_ids.flatten()

        indices = [i for i, (x, y) in enumerate(zip(pred_flat, labels_flat)) if x != y]

    return indices

indices = mis_classified_examples(test_set)

for i in indices:
    print(test_text[i])
    print()

داخل توضیحات نوشته زبان فارسی و انگلیسی ولی زیر نویس فارسی بود ای کاش دوبله فارسی اضافه بشه اوج بدبختیه که بازی سازندش ایرانیه ولی زبان فارسی نداره سازنده حداقل احترام برای مخاطبای فارسی زبان کشورش باید بزاره در کل خوب بود

ویژگی های مثبت: 1-جنگل پویا:به خصوص تو منرحله ماسوله برف ها با کیفیت بودند(برف هاش با کیفیت بودن ولی مه اش با کیفیت تر بودند،انگار داشتم بازیsilent hillبازی میکرم{البته منsilent hillبازی نکردم بلکه فیلم کوتاهی از گیمپلی اش رو دیدم}) 2-صدا بعضی از شخضیت ها با کیفیت بودند3-شاید تیربارهای موجود تو بازی گرافیک با کیفیت ندارند اما صدا واقعا" اکشن صدا می شود. بازی چندان خوبی نبود.اما تو مرحله یکی می دونه خیلی بد بود اصلا فانوس ها رو کجات گذاشتند که من بد از یکم ساعت تونستم پیدا بکنم. سایر ویژگی های منفی: 1-تیر می زنیم به دشمنان اصلا انگار جلیگه ضد گلوله پوشیدند. 2-اصلا تکلیف چیه،بعضی ها از فانوس هال رو باید نزدیک نزدیک بزنیم یا دور د ور بزنیم.و 3-نور پردازی نه چندان خوب در شب 4-شاتگان دو تیر یا همان تفنگ پدر انگار برای دکور با این تفنگ انگار داریم گلوله پلاستیکی به دشمن ش

### *DESCRIBE YOUR QUALITATIVE ANALYSIS OF THE ABOVE EXAMPLES HERE*

Since, our test-set did not have that many reviews/comments, and since, the accuracy of our fine-tuned model was a little bit better, I decided to print out all the reviews/comments the model got the label wrong, so that I could have a better understanding of the performance of the model.

In some reviews/comments, the review/comment comprised of multiple sentences. Some sentences sounded negative, but the overall intent of the user was positive. In these cases, the model could not perform well. Or, the review/comment comprised of some positive sentences, but overall, the user has felt bad about the game. In these cases, which we can see above, the model could not perform very well and accurately label the review/comment. I can claim that the model has not been able to extract the true intent of those reviews/comments and has shown weak performance from itself.

In other cases, the review/comment seems actually easy (like the third review/comment from above), however, the reason as to why the model has not been able to accurately label it, is baffling to me. Maybe the sentence which the user has claimed the game is not bad, made the model think that the review/comment is positive.

Overall, I can conclude that the model is not strong enough to understand the true intent of some reviews/comments, when the reviews/comments comprise of multiple sentences that might be conflicting with one another. In these cases, no matter how many epochs I let the model be trained, it cannot truly understand why a specific review/comment has a specific label. For future work, I suggest using a stronger model, or a model which has had been trained on multiple tasks and bigger dataset in pre-training phase. In this approach, I can claim that the model can perform better. For example, if I feed the reviews/comments to ChatGPT or GPT-4, I would get much better results, even though I cannot fine-tune those models.



---

Finished? Remember to upload the PDF file of this notebook and url of your colab and your three dataset files (annotator1.csv, annotator2.csv, and final_data.csv) to Elearn (elearn.ut.ac.ir)


## AI Disclosure

*   Did you use any AI assistance to complete this homework? If so, please also specify what AI you used.
    * *No*


---
*(only complete the below questions if you answered yes above)*

*   If you used a large language model to assist you, please paste *all* of the prompts that you used below. Add a separate bullet for each prompt, and specify which problem is associated with which prompt.
    * *your response here*
*   **Free response**: For each problem for which you used assistance, describe your overall experience with the AI. How helpful was it? Did it just directly give you a good answer, or did you have to edit it? Was its output ever obviously wrong or irrelevant? Did you use it to get the answer or check your own answer?
    * *your response here*
